File: javascript/016.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var threeSumClosest = function(nums, target) {
  nums.sort(function(a, b) {return a-b;});

  let closestTarget = Number.MAX_VALUE;
  let delta = Number.MAX_VALUE;
  for(let i = 0; i < nums.length - 2; ++i) {
    let j = i + 1;
    let k = nums.length - 1;
    while(j < k) {
      var sum = nums[i] + nums[j] + nums[k];
      if (Math.abs(sum - target) < delta) {
        closestTarget = sum;
        delta = Math.abs(sum - target);
      }
      if (sum > target) {
        --k;
      } else if(sum < target) {
        ++j;
      } else {
        return target;
      }
    }
  }

  return closestTarget;
};

File: ruby/016.rb
# @param {Integer[]} nums
# @param {Integer} target
# @return {Integer}
def three_sum_closest(nums, target)
  nums.sort!

  closest_sum = nil
  i = 0
  while i < nums.length do
    j = i + 1
    k = nums.length - 1
    while j < k do
      sum = nums[i] + nums[j] + nums[k]

      return sum if sum == target

      if closest_sum.nil?
        closest_sum = sum
        next
      end

      if (sum - target).abs < (closest_sum - target).abs
        closest_sum = sum
      end

      if sum > target
        k -= 1
      else
        j += 1
      end
    end

    i += 1
  end
  closest_sum
end